tests/parser/types/test_dynamic_array.py

```diff
@@ -0,0 +1,430 @@
+    return 1,2
+
+@external
+def outer() -> DynArray[int128, 2]:
+    return [333, self.inner()[0]]
+    """
+
+    c = get_contract(code)
+    assert c.outer() == [333, 1]
+
+
+def test_nested_calls_inside_arrays(get_contract):
+    code = """
+@internal
+def _foo(a: uint256, b: DynArray[uint256, 2]) -> (uint256, uint256, uint256, uint256, uint256):
+    return 1, a, b[0], b[1], 5
+
+@internal
+def _foo2() -> uint256:
+    a: DynArray[uint256, 10] = [6,7,8,9,10,11,12,13,15,16]
+    return 4
+
+@external
+def foo() -> (uint256, uint256, uint256, uint256, uint256):
+    return self._foo(2, [3, self._foo2()])
+    """
+
+    c = get_contract(code)
+    assert c.foo() == [1, 2, 3, 4, 5]
+
+
+def test_nested_calls_inside_arrays_with_index_access(get_contract):
+    code = """
+@internal
+def _foo(
+    a: DynArray[uint256, 2],
+    b: DynArray[uint256, 2]
+) -> (uint256, uint256, uint256, uint256, uint256):
+    return a[1]-b[0], 2, a[0]-b[1], 8-b[1], 5
+
+@internal
+def _foo2() -> (uint256, uint256):
+    a: DynArray[uint256, 10] = [6,7,8,9,10,11,12,13,15,16]
+    return a[6], 4
+
+@external
+def foo() -> (uint256, uint256, uint256, uint256, uint256):
+    return self._foo([7, self._foo2()[0]], [11, self._foo2()[1]])
+    """
+
+    c = get_contract(code)
+    assert c.foo() == [1, 2, 3, 4, 5]
+
+
+def test_so_many_things_you_should_never_do(get_contract):
+    code = """
+@internal
+def _foo(a: DynArray[uint256, 2], b: DynArray[uint256, 2]) -> DynArray[uint256, 5]:
+    return [a[1]-b[0], 2, a[0]-b[1], 8-b[1], 5]
+
+@internal
+def _foo2() -> (uint256, uint256):
+    b: DynArray[uint256, 2] = [5, 8]
+    a: DynArray[uint256, 10] = [6,7,8,9,10,11,12,13,self._foo([44,b[0]],b)[4],16]
+    return a[6], 4
+
+@external
+def foo() -> (uint256, DynArray[uint256, 3], DynArray[uint256, 2]):
+    x: DynArray[uint256, 3] = [
+        1,
+        14-self._foo2()[0],
+        self._foo([7,self._foo2()[0]], [11,self._foo2()[1]])[2]
+    ]
+    return 666, x, [88, self._foo2()[0]]
+    """
+    c = get_contract(code)
+    assert c.foo() == [666, [1, 2, 3], [88, 12]]
+
+
+def test_list_of_structs_arg(get_contract):
+    code = """
+struct Foo:
+    x: uint256
+    y: uint256
+
+@external
+def bar(_baz: DynArray[Foo, 3]) -> uint256:
+    sum: uint256 = 0
+    for i in range(3):
+        sum += _baz[i].x * _baz[i].y
+    return sum
+    """
+    c = get_contract(code)
+    c_input = [[x, y] for x, y in zip(range(3), range(3))]
+    assert c.bar(c_input) == 5  # 0 * 0 + 1 * 1 + 2 * 2
+
+
+def test_list_of_structs_arg_with_dynamic_type(get_contract):
+    code = """
+struct Foo:
+    x: uint256
+    _msg: String[32]
+
+@external
+def bar(_baz: DynArray[Foo, 3]) -> String[96]:
+    return concat(_baz[0]._msg, _baz[1]._msg, _baz[2]._msg)
+    """
+    c = get_contract(code)
+    c_input = [[i, msg] for i, msg in enumerate(("Hello ", "world", "!!!!"))]
+    assert c.bar(c_input) == "Hello world!!!!"
+
+
+# TODO CMC 2022-01-02 not sure we should allow constant dyn arrays
```

Comment:
Seems okay to have, e.g. let's say you're working with Uniswap and have a "default" path you'd like to use e.g.:
```py
DEFAULT_PATH = constant(DynArray[Token, 2]) = [DAI, WETH]